### PR TITLE
Added RequestTracker implementation to create support tickets

### PR DESCRIPTION
**Pull request by adaybujeda**
RequestTracker API implementation for support tickets.

It uses the request tracker API to deliver the support ticket.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202821104799596/1203098149705963) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by adaybujeda** [comment]
@johrstrom, this is the Request Tracker API implementation for the support ticket functionality.

It selects between the 2 implementations based on the configuration object.

The default implementation uses the same template.
I am checking with the IQSS team to see if we want to create a custom one. If that is the case, do you want us to contribute the template as well?

**Comment by johrstrom** [comment]
OK - I got another OSC thing come up this week so I be delayed.

File: apps/dashboard/app/apps/request_tracker_client.rb [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  def initialize(config)
+    # DEFAULTS
+    @auth_token = nil
+    @timeout = 30
+    @verify_ssl = false
+    # FROM CONFIGURATION
+    @user = config[:user]
+    @pass = config[:pass]
+    @auth_token = config[:auth_token]
+    @timeout = config[:timeout] if config[:timeout]
```
Comment (johrstrom) [inline apps/dashboard/app/apps/request_tracker_client.rb:44]: Can you do something like this if there are defaults?

```suggestion
    @timeout = config[:timeout] || 30
```

File: apps/dashboard/app/apps/request_tracker_client.rb [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+    options = {
+      headers:    headers,
+      timeout:    @timeout,
+      verify_ssl: @verify_ssl
+    }
+    options[:proxy] = config[:proxy] if config[:proxy]
+
+    @rt_client = RestClient::Resource.new(@resource, options)
+    untaint
```
Comment (johrstrom) [inline apps/dashboard/app/apps/request_tracker_client.rb:74]: Why is this here?

**Comment by johrstrom** [comment]
Can you provide a config I can work off of (obviously without the secret stuff).

Also related to the secret stuff - none of this will actually be secret. I.e., if a user can read it (which they can/will need to) then they've got it.  There any concern on your side for someone potentially using this key maliciously?

File: apps/dashboard/app/apps/request_tracker_client.rb [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+  def create(field_hash)
+    field_hash[:id] = 'ticket/new'
+    payload = compose(field_hash)
+    resp = @rt_client['ticket/new/edit'].post payload
+    new_id = resp.match(/Ticket\s*(\d+)/)
+    if new_id.instance_of?(MatchData)
+      new_id[1]
+    else
+      raise "Unable to create ticket. Server response: #{resp}"
```
[the remark this answers is not in the document]
Comment (adaybujeda) [reply apps/dashboard/app/apps/request_tracker_client.rb]: Updated

File: apps/dashboard/app/apps/request_tracker_client.rb (hunk 1 above)
Comment (adaybujeda) [reply apps/dashboard/app/apps/request_tracker_client.rb]: Updated defaults setup

File: apps/dashboard/app/apps/request_tracker_client.rb (hunk 2 above)
Comment (adaybujeda) [reply apps/dashboard/app/apps/request_tracker_client.rb]: The client is based on the RT client library: https://github.com/uidzip/rt-client
`untaint` is originally in the lib client and I didn't remove or question its use.

I have removed from the code and it works as before.

**Comment by adaybujeda** [comment]
This is my local config for RT:
```yaml
support_ticket:
  rt_api:
    server: "http://rt:34000"
    user: "root"
    pass: "password"
    priority: 4
    queues:
      - "General"
```

We use a RequestTracker Docker image for local development: netsandbox/request-tracker:4.4
`docker run -it --rm --env RT_WEB_PORT=34000 -p 34000:34000 netsandbox/request-tracker:4.4`

Default username and password: root => pasword
The web interface is at: http://localhost:34000

This is our Docker Compose config:
```yaml
  rt:
    image: netsandbox/request-tracker:4.4
    hostname: rt
    container_name: rt
    environment:
      RT_WEB_PORT: "34000"
    ports:
      - "34000:34000"
    expose:
      - "34000"
```

**Comment by adaybujeda** [comment]
> Also related to the secret stuff - none of this will actually be secret. I.e., if a user can read it (which they can/will need to) then they've got it. There any concern on your side for someone potentially using this key maliciously?

We discussed this internally when we implemented the RT integration for the Sid project. We came to the conclusion that credentials with restricted permissions, that can only create tickets, are OK to not be a secret.

**Comment by johrstrom** [comment]
> We discussed this internally when we implemented the RT integration for the Sid project. We came to the conclusion that credentials with restricted permissions, that can only create tickets, are OK to not be a secret.

Cool!

> Would it be more secure to use environment variables for username/password/auth_token?
> We could update to use ENV[] first and use config when ENV is not defined.

I think config is fine. A sophisticated user is going to find it either way, so an ENV variable isn't really more secure because it's just a different config file to look at.

**Comment by adaybujeda** [comment]
@johrstrom any updates on this PR?

**Comment by johrstrom** [comment]
Hi sorry I was at the gateways conference all last week. We'll close this out this week.

**Review by johrstrom — APPROVED** [review]
This looks fine by me, there are a couple things we could refactor like how we render the ERB, but with so much going on this is fine.

**Comment by adaybujeda** [comment]
Thanks Jeff!! 🚀
